server/aesdsocket.c

```diff
@@ -0,0 +1,477 @@
+                    }
+
+                    file_size += index;
+
+                    //Send all the contents read from /var/tmp/aesdsocketdata back to the client
+                    if(lseek(file_fd, 0, SEEK_SET) == -1)
+                    {
+                        syslog(LOG_ERR, "Could not get to the beginning of the file: %s", strerror(errno));
+                        exit(1);  
+                    }
+                    
+                    //Perform reads to send the file contents to the socket client
+                    int to_be_sent = file_size;
+                    char buff_read[RECV_BUFF_LEN];
+                    while(to_be_sent)
+                    {
+                        int send_bytes = 0;
+                        int read_bytes = read(file_fd, buff_read, RECV_BUFF_LEN);
+                        if(read_bytes != 0)
+                            send_bytes = read_bytes;
+
+                        if(read_bytes == -1)
+                        {
+                            //If the error is caused by an interruption of the system call try again
+                            if(errno == EINTR)
+                                continue;
+
+                            //Else, error occurred, print it to syslog and finish program
+                            syslog(LOG_ERR, "Could not read from the file: %s", strerror(errno));
+                            exit(1);
+                        }
+
+                        //Less bytes remaining
+                        to_be_sent -= read_bytes;
+                        
+                        //Send the contents back to the client
+                        int sent_bytes = -1;
+                        int send_off = 0;
+                        syslog(LOG_INFO, "Send bytes: %d", send_bytes);
+                        while(sent_bytes != 0)
+                        {
+                            sent_bytes = send(connection_fd, &buff_read[send_off], send_bytes, 0);
+                            if(sent_bytes == -1)
+                            {
+                                //If the error is caused by an interruption of the system call try again
+                                if(errno == EINTR)
+                                    continue;
+
+                                //Else, error occurred, print it to syslog and finish program
+                                syslog(LOG_ERR, "Could not read from the file: %s", strerror(errno));
+                                exit(1);
+                            }
+                            send_bytes -= sent_bytes;
+                            send_off += sent_bytes;
+                        }
+                        
+                    }
+
+                    //Reset index to use the malloc'ed buffer from the beginning
+                    index = 0;
+                }
+                //Realloc the array if it got full without an '\n'
+                else if(index == (RECV_BUFF_LEN*chunks))
```

Comment:
@jordicrosmompart I think this logic has the potential to exceed array bounds.  Consider the case where a small buffer of bytes was received by `recv` and did not include a packet terminator.  Since `index == (RECV_BUFF_LEN*chunks)` would not be true, you'd go back to the `recv` call and `recv(connection_fd, &recv_data[index], RECV_BUFF_LEN, 0);`.  However, index would be non-zero, and index+RECV_BUFF_LEN would exceed array bounds of the buffer sized `RECV_BUFF_LEN` bytes.